logicalplan/merge_selects.go

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) The Thanos Community Authors.
+// Licensed under the Apache License 2.0.
+
+package logicalplan
+
+import (
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql/parser"
+)
+
+// MergeSelectsOptimizer optimizes a binary expression where
```

Comment:
So it feels like this optimizer is more of a cache-based optimization rather than a rule-based one? I think we can also implement this in the physical plan directly as an optimization effort, instead of extending PromQL and having a separate `filter()` for it as a rule in the logical plan. It feels like this doesn't really "rewrite the query", but rather implements something new.

We definitely should do this caching, but I wonder if the logical plan is the right place to implement it.
This is new to me, so feel free to correct me if I'm getting this wrong somewhere! 🙂

Reply:
Yeah that's true, we are adding two optimizations in this case. The physical plan can cache series calls, and logical optimizers make sure that the cache-hit ratio is optimized. 

I think adding optimizers to the physical plan would not be straightforward, since the phyisical plan is essentially a set of instantiated operators. We would need to figure out how to transform and rewire them.

Reply:
Let's keep it simple -> optimizer changing logical plan

Reply:
Optimizers should only optimize the logical plan indeed. The physical plan is then constructed from that optimized logical plan and can make certain assumptions while being constructed.